### PR TITLE
Enhance mask_to_segs to Handle Batch Masks for Improved Facial Reconstruction Workflow

### PR DESCRIPTION
**Pull request by lcolok**
Hello there,

First and foremost, I would like to express my deepest gratitude for maintaining such an excellent project. It has provided immense help in my research and development efforts. The SAM and Detailer nodes offered by your project have been particularly beneficial to me, and the thoughtful design of the Pipe class nodes, especially when paired with your workflow component plugin, is an efficient and reusable tool.

The origin of this PR comes from the construction of a facial repair workflow where the output generated from converting the Combined Mask into SEGS was less than ideal. The logic of Mask To Segs for single images infers based on the closed areas in the mask, a "lossy process". When there is overlap between the borders of two masks, they get recognized as one SEG region. As shown in the example below, the two characters (parents) have heads that are too close together, resulting in a single SEG region. Therefore, when the SD is called for img2img processing in the Detailer node, it processes them together rather than separately. Strangely, an erroneous partition was created that flattened the characters' faces.

image

To resolve this issue, I have expanded the function to process batch masks, maintaining the independence of each mask without the need for inference. This is a lossless process and, as the results indicate, a substantial improvement.

image

image

Here are the key improvements to the mask_to_segs function:

- Added support for different input types (NumPy arrays or tensors).
- Implemented dimension checks and adjustments for mask inputs, allowing support for both single and batch images.
- Introduced a batch processing loop for handling multiple mask images.

This update should enable a more flexible and robust facial reconstruction workflow, especially when dealing with overlapping masks. I believe this enhancement will be valuable to others in the community as well.

I look forward to any feedback you may have, and I hope that my contribution can be of help to this impressive project.

Thank you for your time and consideration.

### CODE REVIEW

**Comment by ltdrdata** [comment]
Nice work. I'll test it out when I get home and then merge it.

**Comment by ltdrdata** [comment]
I found a bug from here. [clipseg](https://github.com/ltdrdata/ComfyUI-extension-tutorials/raw/Main/ComfyUI-Impact-Pack/workflow/clipseg.json)

https://github.com/ltdrdata/ComfyUI-extension-tutorials/tree/Main/ComfyUI-Impact-Pack/workflow

Maybe you can reproduce the symptoms without my checkpoint from this image.
ComfyUI_00126_

```
  File "C:\Users\rho\git\ComfyUI\execution.py", line 145, in recursive_execute
    output_data, output_ui = get_output_data(obj, input_data_all)
  File "C:\Users\rho\git\ComfyUI\execution.py", line 75, in get_output_data
    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
  File "C:\Users\rho\git\ComfyUI\execution.py", line 68, in map_node_over_list
    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
  File "C:\Users\rho\git\ComfyUI-Impact-Pack\modules\impact\impact_pack.py", line 1337, in doit
    enhanced_img, cropped_enhanced, mask = FaceDetailer.enhance_face(
  File "C:\Users\rho\git\ComfyUI-Impact-Pack\modules\impact\impact_pack.py", line 801, in enhance_face
    mask = core.segs_to_combined_mask(segs)
  File "C:\Users\rho\git\ComfyUI-Impact-Pack\modules\impact\core.py", line 714, in segs_to_combined_mask
    mask[crop_region[1]:crop_region[3], crop_region[0]:crop_region[2]] |= (cropped_mask * 255).astype(np.uint8)
ValueError: non-broadcastable output operand with shape (1,357) doesn't match the broadcast shape (453,357)
```

**Comment by lcolok** [comment]
It's a bit strange. I haven't modified the function segs_to_combined_mask. Let me carefully review it after I finish installing the CLIPSeg script.

**Comment by lcolok** [comment]
In the function segs_to_combined_mask, I printed the shape and the result was `(1, 512, 360)`. In the subsequent two lines of code, I assigned the width and height values as follows:
```
h = shape[0]
w = shape[1]
```
The problem became apparent - the height should correspond to the element at index 2, and the width should correspond to the element at index 3. Therefore, the solution is to modify the line:
```
shape = segs[0]
```
to:
```
shape = segs[0][1:]
```

However, this raises a deeper question: why is the shape tuple passed with three elements? Does the value 1 represent the batch dimension or the channel dimension in the (C, H, W) format? I believe it represents the channel dimension. Recently, when I was writing custom nodes, I noticed that in ComfyUI, the shape of a regular image is in the (H, W, C) format, indicating channel-last ordering. However, the shape used to represent masks is (C, H, W). Has there been a change in conventions after a certain version?

**Comment by ltdrdata** [comment]
> In the function segs_to_combined_mask, I printed the shape and the result was `(1, 512, 360)`. In the subsequent two lines of code, I assigned the width and height values as follows:
> 
> ```
> h = shape[0]
> w = shape[1]
> ```
> 
> The problem became apparent - the height should correspond to the element at index 2, and the width should correspond to the element at index 3. Therefore, the solution is to modify the line:
> 
> ```
> shape = segs[0]
> ```
> 
> to:
> 
> ```
> shape = segs[0][1:]
> ```
> 
>  However, this raises a deeper question: why is the shape tuple passed with three elements? Does the value 1 represent the batch dimension or the channel dimension in the (C, H, W) format? I believe it represents the channel dimension. Recently, when I was writing custom nodes, I noticed that in ComfyUI, the shape of a regular image is in the (H, W, C) format, indicating channel-last ordering. However, the shape used to represent masks is (C, H, W). Has there been a change in conventions after a certain version?

I am confused always, too. lol...

mask, latent, image, ...  they all have different shape of dimensions.
When you digging inside the code. You can find some code uses (b, c, h, w).
But the latent is (b, h, w, c).

We need developer documentation.

**Comment by lcolok** [comment]
> > In the function segs_to_combined_mask, I printed the shape and the result was `(1, 512, 360)`. In the subsequent two lines of code, I assigned the width and height values as follows:
> > ```
> > h = shape[0]
> > w = shape[1]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The problem became apparent - the height should correspond to the element at index 2, and the width should correspond to the element at index 3. Therefore, the solution is to modify the line:
> > ```
> > shape = segs[0]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to:
> > ```
> > shape = segs[0][1:]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> >  However, this raises a deeper question: why is the shape tuple passed with three elements? Does the value 1 represent the batch dimension or the channel dimension in the (C, H, W) format? I believe it represents the channel dimension. Recently, when I was writing custom nodes, I noticed that in ComfyUI, the shape of a regular image is in the (H, W, C) format, indicating channel-last ordering. However, the shape used to represent masks is (C, H, W). Has there been a change in conventions after a certain version?
> 
> I am confused always, too. lol...
> 
> mask, latent, image, ... they all have different shape of dimensions. When you digging inside the code. You can find some code uses (b, c, h, w). But the latent is (b, h, w, c).
> 
> We need developer documentation.

Indeed, it's highly necessary. Fortunately, with your Impact Pack and the WAS custom node project, I've saved a lot of time fumbling in the dark with ComfyUI by learning many abstracted functions and utilities from both of you. Now I have two torches illuminating my path: PyTorch and your project.

**Comment by lcolok** [comment]
I'm sorry, I made a mistake and accidentally submitted some duplicate commits. However, I don't know how to undo them.

**Comment by ltdrdata** [comment]
nvm. I'll squash.

File: modules/impact/impact_pack.py [hunk 1]
```diff
@@ -419,14 +421,25 @@ def do_detail(image, segs, model, clip, vae, guide_size, guide_size_for, max_siz
...
             cropped_list.append(torch.from_numpy(cropped_image))
 
         image_tensor = pil2tensor(image_pil.convert('RGB'))
 
         cropped_list.sort(key=lambda x: x.shape, reverse=True)
         enhanced_list.sort(key=lambda x: x.shape, reverse=True)
+        enhanced_alpha_list.sort(key=lambda x: x.shape, reverse=True)
 
-        return image_tensor, cropped_list, enhanced_list
+        return image_tensor, cropped_list, enhanced_list, enhanced_alpha_list
```
Comment (ltdrdata) [inline modules/impact/impact_pack.py:438]: This modification affects multiple call sites. You need check that.
DetailerForEach, DetailerForEachTest, FaceDetailer,  ...

You can not execute clipseg.json successfully.

**Review by ltdrdata — APPROVED** [review]
masks
[test/masks.json](https://github.com/ltdrdata/ComfyUI-Impact-Pack/raw/41ae5296c962e3545ef1ffe6978e3dede36aa75e/test/masks.json)

Assuming that MASK is not inherently treated as a batch, handling it in a form that can be batched can lead to various errors. To clarify, I have separated the batched masks as MASKS.

Additionally, to ensure compatibility with external nodes, I have added a separate node that can convert Masks to Mask List.

Verify if this form aligns with the use case you proposed.

During the testing process, I noticed that the current grouping method, where segments are grouped into sets of three using an arbitrary group size, has a significant level of randomness. I believe there is a need for improvement in the future, particularly by considering factors such as overlap, adjacency, and size of the masks. By incorporating these considerations, the grouping process can be enhanced to provide more consistent and meaningful results.

I will first verify the proposed changes, and if I believe they align with your suggestions, I will proceed with merging them.

**Comment by lcolok** [comment]
> masks
> 
> [test/masks.json](https://github.com/ltdrdata/ComfyUI-Impact-Pack/raw/41ae5296c962e3545ef1ffe6978e3dede36aa75e/test/masks.json)
> 
> 
> 
> Assuming that MASK is not inherently treated as a batch, handling it in a form that can be batched can lead to various errors. To clarify, I have separated the batched masks as MASKS.
> 
> 
> 
> Additionally, to ensure compatibility with external nodes, I have added a separate node that can convert Masks to Mask List.
> 
> 
> 
> Verify if this form aligns with the use case you proposed.
> 
> 
> 
> During the testing process, I noticed that the current grouping method, where segments are grouped into sets of three using an arbitrary group size, has a significant level of randomness. I believe there is a need for improvement in the future, particularly by considering factors such as overlap, adjacency, and size of the masks. By incorporating these considerations, the grouping process can be enhanced to provide more consistent and meaningful results.
> 
> 
> 
> I will first verify the proposed changes, and if I believe they align with your suggestions, I will proceed with merging them.
> 
> 
> 
> 

I think it's working great now! However, it's important to note that converting a single mask into multiple smaller region masks is a lossy process, as you mentioned: overlapping and adjacency issues are unavoidable. So, the most accurate approach is indeed to input a mask list. To address this, we can explore other methods such as using CLIP Seg for reasoning with multiple keywords or manually employing multiple layers for masking purposes.

**Pull request merged**